### PR TITLE
[JIT] Backport: 8331087: Move immutable nmethod data from CodeCache

### PR DESCRIPTION
**Pull request by kuaiwei**
Summary: Add option ReduceNMethodSize. If enable, jvm will move
         immutable nmethod data out of CodeCache

Testing: CI testing

Reviewers: yueshi.zwj, lusou.zq

Issue: https://github.com/dragonwell-project/dragonwell11/issues/842

### CODE REVIEW

File: src/hotspot/share/code/nmethod.hpp [hunk 1]
```diff
@@ -58,9 +58,12 @@ class nmethod : public CompiledMethod {
   friend class NMethodSweeper;
   friend class CodeCache;  // scavengable oops
  private:
+  // nmethod's read-only data
+  address _immutable_data;
 
   // Shared fields for all nmethod's
   int       _entry_bci;        // != InvocationEntryBci if this nmethod is an on-stack replacement method
+  int       _immutable_data_size;
```
Comment (lusou-zhangquan) [inline src/hotspot/share/code/nmethod.hpp:66]: It seems that field _immutable_data_size and _immutable_data are not initialized in constructor for native wrappers.

Comment (kuaiwei) [reply src/hotspot/share/code/nmethod.hpp:66]: Thanks for catch it.

**Review by lusou-zhangquan — APPROVED** [review]
LGTM

File: src/hotspot/share/code/nmethod.hpp [hunk 2]
```diff
@@ -281,23 +289,89 @@ class nmethod : public CompiledMethod {
...
+    } else {
+      return header_begin() + _scopes_pcs_offset;
+    }
+  }
+  PcDesc* scopes_pcs_begin() const {
+    if (_immutable_data_size > 0) {
+      return (PcDesc *) (immutable_data_begin() + _scopes_pcs_offset);
+    } else {
+      return (PcDesc *) (header_begin() + _scopes_pcs_offset);
+    }
```
Comment (JoshuaZhuwj) [inline src/hotspot/share/code/nmethod.hpp:321]: Just a small nit.
PcDesc* scopes_pcs_begin() const {
  return (PcDesc *) scopes_data_end();
}

File: src/hotspot/share/code/nmethod.hpp [hunk 3]
```diff
@@ -281,23 +289,89 @@ class nmethod : public CompiledMethod {
...
+  }
+  PcDesc* scopes_pcs_end() const {
+    if (_immutable_data_size > 0) {
+      return (PcDesc *) (immutable_data_begin() + _dependencies_offset);
+    } else {
+      return (PcDesc *) (header_begin() + _dependencies_offset);
+    }
+  }
```
Comment (JoshuaZhuwj) [inline src/hotspot/share/code/nmethod.hpp:329]: Same with the above comment.
return (PcDesc *) dependencies_begin();

And also some of the following codes could be reusable.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/NMethod.java [hunk 4]
```diff
@@ -136,16 +142,76 @@ public Address getAddress() {
...
+    }
+  }
+  public Address scopesDataEnd()        {
+    if (getImmutableDataSize() > 0) {
+      return immutableDataBegin().addOffsetTo(getScopesPCsOffset());
+    } else {
+      return headerBegin().addOffsetTo(getScopesPCsOffset());
+    }
+  }
+  public Address scopesPCsBegin()       {
```
Comment (JoshuaZhuwj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/NMethod.java:159]: Same with the above comment.

File: src/hotspot/share/code/nmethod.hpp (hunk 2 above)
Comment (kuaiwei) [reply src/hotspot/share/code/nmethod.hpp:321]: It follows the origin code. like
```c++
  address scopes_data_end       () const          { return           header_begin() + _scopes_pcs_offset    ; }
  PcDesc* scopes_pcs_begin      () const          { return (PcDesc*)(header_begin() + _scopes_pcs_offset   ); }
```
The origin code doesn't reuse. I just follow it.

**Review by JoshuaZhuwj — APPROVED** [review]

**Pull request merged**